src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -514,27 +478,79 @@ else if (null != mHeader) {
+        Integer referenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
+        if (!NO_ALIGNMENT_REFERENCE_NAME.equals(referenceName)) {
+            if (null == header) {
+                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the reference index or name");
+            }
+            referenceIndex = header.getSequenceIndex(referenceName);
+            if (NO_ALIGNMENT_REFERENCE_INDEX == referenceIndex) {
+                if (strict) {
+                    throw new IllegalArgumentException("Reference index for '" + referenceName + "' not found in sequence dictionary.");
+                }
+                else {
+                    referenceIndex = null;  // unresolved.
+                }
+            }
         }
-        else {
-            SAMSequenceRecord samSequence = mHeader.getSequence(mateReferenceIndex);
-            if (null != samSequence) {
-                mMateReferenceIndex = mateReferenceIndex;
-                mMateReferenceName = samSequence.getSequenceName();
+        return referenceIndex;
+    }
+
+    /**
+     * Static method that resolves and returns the reference name corresponding to a given reference index.
+     *
+     * @param referenceIndex If {@code referenceIndex} is NO_ALIGNMENT_REFERENCE_INDEX, the value NO_ALIGNMENT_REFERENCE_NAME
+     *                      is returned directly. Otherwise {@code referenceIndex} must be looked up in the header's sequence
+     *                      dictionary.
+     * @param header SAMFileHeader to use when resolving {@code referenceIndex} to a name. Must be non null unless the
+     *               the {@code referenceIndex} is NO_ALIGNMENT_REFERENCE_INDEX.
+     * @returns the reference name corresponding to  {@code referenceIndex}
+     * @throws IllegalStateException if {@code referenceIndex} is not equal to NO_ALIGNMENT_REFERENCE_NAME and the header
+     * is null
+     * @throws IllegalArgumentException if {@code referenceIndex} does not appear in header's sequence dictionary.
+     *
+     * Does not mutate the SAMRecord.
+     */
+    protected static String resolveNameFromIndex(final int referenceIndex, final SAMFileHeader header) {
+        String referenceName = NO_ALIGNMENT_REFERENCE_NAME;
+        if (NO_ALIGNMENT_REFERENCE_INDEX != referenceIndex) {
+            if (null == header) {
+                throw new IllegalStateException("A non-null SAMFileHeader is required to resolve the reference index or name");
             }
-            else {
-                throw new IllegalArgumentException("Reference index " + mateReferenceIndex + " not found in sequence dictionary.");
+            SAMSequenceRecord samSeq = header.getSequence(referenceIndex);
+            if (null == samSeq) {
+                throw new IllegalArgumentException("Reference name for '" + referenceIndex + "' not found in sequence dictionary.");
```

Comment:
Right - I originally had a strict mode for this as well but the policy was and is that you can't set an unresolveable index on a record, only an unresolveable name.